src/utils/codec.ts

```diff
@@ -0,0 +1,318 @@
+
+/**
+ * Hex Encoding
+ */
+// prettier-ignore
+export const HEX: CodecScheme = {
+    bits: 4,
+    table: [
+        "0", "1", "2", "3",
+        "4", "5", "6", "7",
+        "8", "9", "a", "b",
+        "c", "d", "e", "f",
+    ],
+    tail: "",
+}
+
+/**
+ * Base64 Encoding
+ */
+// prettier-ignore
+export const BASE64: CodecScheme = {
+    bits: 6,
+    table: [
+        "A", "B", "C", "D", "E", "F",
+        "G", "H", "I", "J", "K", "L",
+        "M", "N", "O", "P", "Q", "R",
+        "S", "T", "U", "V", "W", "X",
+        "Y", "Z", "a", "b", "c", "d",
+        "e", "f", "g", "h", "i", "j",
+        "k", "l", "m", "n", "o", "p",
+        "q", "r", "s", "t", "u", "v",
+        "w", "x", "y", "z", "0", "1",
+        "2", "3", "4", "5", "6", "7",
+        "8", "9", "+", "/"
+    ],
+    tail: "=",
+}
+
+/**
+ * Unicode code points range: `0 - 0x10FFFF`.
+ *
+ * If there is no element at point, the value is `undefined`.
+ */
+type Point = number | undefined
+
+/**
+ * Codec
+ *
+ * Encoding/Decoding string with Unicode code points(0 - 0x10FFFF).
+ *
+ * All data will be formated into binary while constructing Codec
+ * class, for example, Let's format the word `Mask`:
+ *
+ * For `Base1024`:
+ *
+ * | Text |   Binary   | Index |
+ * | ---- | ---------- | ----- |
+ * |      | 0100110101 |  309  |
+ * |      | 1000010111 |  535  |
+ * |      | 0011011011 |  219  |
+ *
+ * For `ASCII`: `[ 01001101, 01100001, 01110011, 01101011 ]`
+ *
+ * | Text |  Binary  | Index |
+ * | ---- | -------- | ----- |
+ * |  M   | 01001101 |  77   |
+ * |  a   | 01100001 |  97   |
+ * |  s   | 01110011 |  115  |
+ * |  k   | 01101011 |  107  |
+ *
+ * For `Base64`: `[ 010011, 010110, 000101, 110011, 011010, 11 ]`
+ *
+ * | Text | Binary | Index |
+ * | ---- | ------ | ----- |
+ * |  T   | 010011 |  19   |
+ * |  W   | 010110 |  22   |
+ * |  F   | 000101 |  5    |
+ * |  z   | 110011 |  51   |
+ * |  a   | 011010 |  26   |
+ * |  s   | 110000 |  48   |
+ * |  =   | 000000 |       |
+ *
+ * For `hex`: `[ 0100, 1101, 0110, 0001, 0111, 0011, 0110, 1011 ]`
+ *
+ * | Text | Binary | Index |
+ * | ---- | ------ | ----- |
+ * |  4   |  0100  |   4   |
+ * |  d   |  1101  |   13  |
+ * |  6   |  0110  |   6   |
+ * |  1   |  0001  |   1   |
+ * |  7   |  0111  |   7   |
+ * |  3   |  0011  |   3   |
+ * |  6   |  0110  |   6   |
+ * |  b   |  1011  |   11  |
+ *
+ * @property {Point[]} points - Code points
+ *
+ * @method  toHex            - Convert code points to hex string
+ * @method  toBase64         - Convert code points to base64 string
+ * @method  toString         - Convert code points to utf-16 string
+ */
+export default class Codec {
+    /**
+     * Build `Codec` from target scheme string
+     *
+     * @param {CodecScheme} scheme - target codec scheme
+     * @param {string}      str    - the base `CodecScheme` string
+     */
+    static from(scheme: CodecScheme, str: string): Codec {
+        return new Codec(
+            Array.from(str, (byte: string) => {
```

Comment:
plz extract map function to a single function.

Reply:
Single function with a new name or zip all code?

Reply:
```typescript
const mapFn = (point: string) => point
return new Codec(Array.from(input, mapFn).join(''))
```